test/validation/api/ipsec/ipsec_test_in.c

```diff
@@ -1727,6 +1768,263 @@ static void test_ipsec_sa_print(void)
 	ipsec_sa_destroy(in_sa);
 }
 
+static void test_in_ipv4_esp_reass_success_two_frags(odp_ipsec_sa_t out_sa,
+						     odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+	int i;
+
+	memset(test_in, 0, sizeof(test_in));
+
+	CU_ASSERT(MAX_FRAGS >= 2);
+
+	part_prep_esp(test_out, 2, false);
+
+	test_out[0].pkt_in = &pkt_ipv4_udp_p1_f1;
+	test_out[1].pkt_in = &pkt_ipv4_udp_p1_f2;
+
+	part_prep_plain(&test_in[1], 1, 1, false, true);
+	test_in[1].out[0].pkt_res = &pkt_ipv4_udp_p1;
+
+	for (i = 0; i < 2; i++) {
+		ipsec_test_packet test_pkt;
+		odp_packet_t pkt;
+
+		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
+
+		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
+		test_in[i].pkt_in = &test_pkt;
+
+		ipsec_check_in_one(&test_in[i], in_sa);
+	}
+}
+
+static void test_in_ipv4_esp_reass_success_four_frags(odp_ipsec_sa_t out_sa,
+						      odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+	int i;
+
+	memset(test_in, 0, sizeof(test_in));
+
+	CU_ASSERT(MAX_FRAGS >= 4);
+
+	part_prep_esp(test_out, 4, false);
+
+	test_out[0].pkt_in = &pkt_ipv4_udp_p2_f1;
+	test_out[1].pkt_in = &pkt_ipv4_udp_p2_f2;
+	test_out[2].pkt_in = &pkt_ipv4_udp_p2_f3;
+	test_out[3].pkt_in = &pkt_ipv4_udp_p2_f4;
+
+	part_prep_plain(&test_in[3], 1, 1, false, true);
+	test_in[3].out[0].pkt_res = &pkt_ipv4_udp_p2;
+
+	for (i = 0; i < 4; i++) {
+		ipsec_test_packet test_pkt;
+		odp_packet_t pkt;
+
+		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
+
+		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
+		test_in[i].pkt_in = &test_pkt;
+
+		ipsec_check_in_one(&test_in[i], in_sa);
+	}
+}
+
+static void test_in_ipv4_esp_reass_success_two_frags_ooo(odp_ipsec_sa_t out_sa,
+							 odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+	int i;
+
+	memset(test_in, 0, sizeof(test_in));
+
+	CU_ASSERT(MAX_FRAGS >= 2);
+
+	part_prep_esp(test_out, 2, false);
+
+	test_out[0].pkt_in = &pkt_ipv4_udp_p1_f2;
+	test_out[1].pkt_in = &pkt_ipv4_udp_p1_f1;
+
+	part_prep_plain(&test_in[1], 1, 1, false, true);
+	test_in[1].out[0].pkt_res = &pkt_ipv4_udp_p1;
+
+	for (i = 0; i < 2; i++) {
+		ipsec_test_packet test_pkt;
+		odp_packet_t pkt;
+
+		CU_ASSERT_EQUAL(ipsec_check_out(&test_out[i], out_sa, &pkt), 1);
+
+		ipsec_test_packet_from_pkt(&test_pkt, &pkt);
+		test_in[i].pkt_in = &test_pkt;
+
+		ipsec_check_in_one(&test_in[i], in_sa);
+	}
+}
+
+static void test_in_ipv4_esp_reass_success_four_frags_ooo(odp_ipsec_sa_t out_sa,
+							  odp_ipsec_sa_t in_sa)
+{
+	ipsec_test_part test_out[MAX_FRAGS], test_in[MAX_FRAGS];
+	int i;
+
+	memset(test_in, 0, sizeof(test_in));
+
+	CU_ASSERT(MAX_FRAGS >= 4);
+
+	part_prep_esp(test_out, 4, false);
+
+	test_out[0].pkt_in = &pkt_ipv4_udp_p2_f1;
+	test_out[1].pkt_in = &pkt_ipv4_udp_p2_f2;
+	test_out[2].pkt_in = &pkt_ipv4_udp_p2_f3;
+	test_out[3].pkt_in = &pkt_ipv4_udp_p2_f4;
```

Comment:
The fragments here should be out of order. Now this function is exactly the same as test_in_ipv4_esp_reass_success_four_frags().

Reply:
Missed this. Will address in next version.